### PR TITLE
User registration system

### PR DESCRIPTION
**Pull request by westlywright**
The Mysql dump is below for the schema as are the env keys im using in the new api file 

### Keys
```
SENDGRID_API_KEY
CATTLE_ACCESS_KEY
CATTLE_SECRET_KEY
CREATE_USER_TEMPLATE_ID
PASSWORD_RESET_TEMPLATE_ID
PASSWORD_VERIFY_RESET_TEMPLATE_ID
```

### Dump
```
-- MySQL dump 10.13  Distrib 5.7.17, for Linux (x86_64)
--
-- Host: db    Database: ui
-- ------------------------------------------------------
-- Server version	5.7.17

/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8 */;
/*!40103 SET @OLD_TIME_ZONE=@@TIME_ZONE */;
/*!40103 SET TIME_ZONE='+00:00' */;
/*!40014 SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0 */;
/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;

--
-- Table structure for table `challenge`
--

DROP TABLE IF EXISTS `challenge`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `challenge` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `account_id` varchar(255) DEFAULT NULL,
  `name` varchar(255) DEFAULT NULL,
  `email` varchar(255) DEFAULT NULL,
  `token` varchar(255) DEFAULT NULL,
  `request` varchar(255) DEFAULT NULL,
  `created` datetime DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`),
  UNIQUE KEY `token_uniq` (`token`)
) ENGINE=InnoDB AUTO_INCREMENT=7 DEFAULT CHARSET=latin1;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Dumping data for table `challenge`
--

LOCK TABLES `challenge` WRITE;
/*!40000 ALTER TABLE `challenge` DISABLE KEYS */;
/*!40000 ALTER TABLE `challenge` ENABLE KEYS */;
UNLOCK TABLES;
/*!40103 SET TIME_ZONE=@OLD_TIME_ZONE */;

/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
/*!40014 SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS */;
/*!40014 SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS */;
/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
/*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
/*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;
/*!40111 SET SQL_NOTES=@OLD_SQL_NOTES */;```

### CODE REVIEW

File: server/caas/api.js [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+  var config = require('../../config/environment')().APP;
+  var request = require('request');
+
+  var helper = require('sendgrid').mail;
+  var crypto = require('crypto');
+
+
+  app.use(bodyParser.json()); // for parsing application/json
+
+  var pool = mysql.createPool({
```
Comment (vincent99) [inline server/caas/api.js]: Need to get all these for ENV so I can set them

File: server/caas/api.js [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        }
+      }
+    });
+  });
+
+  app.use('/reset-password', function(req, res, next) {
+    var user = req.body;
+    var userEmail = user.email;
+    var name = user.name;
+    var url = config.apiServer + config.apiEndpoint + '/credentials?kind=password&publicValue=' + userEmail + '&limit=-1&sort=name';
```
Comment (vincent99) [inline server/caas/api.js]: Don't trust user-input.  Bail if userEmail is empty, encodeUriComponent()

File: server/caas/api.js [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        var credId = body.data[0].id;
+        var url = config.apiServer + config.apiEndpoint + '/accounts/' + body.data[0].accountId;
+
+        newRequest({
+          url: url,
+          method: 'GET'
+        }, function(response, body) {
+
+          if (body.type === 'account') {
+            generateAndInsertToken(body.id, name, userEmail, "reset", function(error, results, fields, token) {
```
Comment (vincent99) [inline server/caas/api.js]: Use the email address from the credential record, not the one the user provided

File: server/caas/api.js [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+      }
+    });
+  });
+
+  app.use('/update-password', function(req, res, next) {
+    var user = req.body;
+    var url = config.apiServer + config.apiEndpoint;
+
+    pool.query('SELECT * FROM challenge WHERE token = ? AND created > DATE_SUB(NOW(), INTERVAL 24 HOUR)', [user.token], function(error, results, fields) {
+      if (error) res.status(500).send({error: 'There was an error finding the record'});
```
Comment (vincent99) [inline server/caas/api.js]: This should say something like the token is invalid or expired, please try requesting a new reset.

File: server/caas/api.js [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+
+      if (results.length) {
+        var credential = results[0];
+
+        newRequest({
+          url: url + '/passwords?publicValue=' + credential.email,
```
Comment (vincent99) [inline server/caas/api.js]: Bail if empty, encodeUriComponent()

File: server/caas/api.js [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+          method: 'GET'
+        }, function(response, body) {
+
+          newRequest({
+            url: url + '/passwords/' + body.data[0].id + '?action=changesecret',
```
Comment (vincent99) [inline server/caas/api.js]: Check body.data to make sure there is one before trying to use it

File: server/caas/api.js [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+
+            newRequest({
+              url: url + '/passwords',
+              method: 'POST',
+              body: credential
+            }, function(response, body) {
+              credentialModel = body;
+              removeUserFromTable(user.email, function(results, fields) {
+                getTokenForLogin(user.email, user.pw, function(response, body) {
+                  res.cookie('token', body.jwt, {secure: true});
```
Comment (vincent99) [inline server/caas/api.js]: `secure: req.secure`

File: server/caas/api.js [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+                  res.redirect('/');
```
Comment (vincent99) [inline server/caas/api.js]: AJAX isn't going to follow this anyway so you might as well just return 204 or 200 and some sort of body.

File: server/caas/api.js (hunk 2 above)
Comment (vincent99) [inline server/caas/api.js]: Bail if empty, encodeUriComponent()

File: server/caas/api.js [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    var userEmail = user.email;
+    var name = user.name;
+    var url = config.apiServer + config.apiEndpoint + '/credentials?kind=password&publicValue=' + userEmail + '&limit=-1&sort=name';
+
+    newRequest({
+      url: url,
+      method: 'GET',
+    }, function(response, body) {
+
+      if (body.data.length) {
```
Comment (vincent99) [inline server/caas/api.js]: Check if body.data exists, if the response is an error you won't have one.

File: server/caas/api.js (hunk 3 above)
Comment (vincent99) [inline server/caas/api.js]: Use email from credential, not user input

File: server/caas/api.js [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+        res.status(500).send({error: 'Credential not found'});
+      }
+    });
+  });
+
+  app.use('/update-password', function(req, res, next) {
+    var user = req.body;
+    var url = config.apiServer + config.apiEndpoint;
+
+    pool.query('SELECT * FROM challenge WHERE token = ? AND created > DATE_SUB(NOW(), INTERVAL 24 HOUR)', [user.token], function(error, results, fields) {
```
Comment (vincent99) [inline server/caas/api.js]: This is used in several places and should just be a helper function, `getChallenge(token_id)`

File: server/caas/api.js (hunk 5 above)
Comment (vincent99) [inline server/caas/api.js]: Bail if empty, encodeUriComponent

File: server/caas/api.js [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+            method: 'POST',
+            body: {newSecret: user.pw, oldSecret: ''}
+          }, function(response, body) {
+            removeUserFromTable(credential.email, function(results, fields) {
+              getTokenForLogin(credential.email, user.pw, function(response, body) {
+                sendPasswordVerificationEmail(credential.email, credential.name, req.headers.origin, function(error, resp) {
+                  if (error) {
+                    res.status(500).send({error: 'error emailing verification'});
+                  } else {
+                    res.cookie('token', body.jwt, {secure: true});
```
Comment (vincent99) [inline server/caas/api.js]: secure: req.secure, consider combining with the same thing earlier so there's one piece of code responsible for making the user logged in.

File: server/caas/api.js [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+      if (response.statusCode >= 200 && response.statusCode < 300) {
+        cb(response, body);
+      } else {
+        var errOut = null;
+        if (error) {
+          errOut = error;
+        } else {
+          errOut = response;
+        }
+        ogRes.status(errOut.statusCode).send({error: errOut});
```
Comment (vincent99) [inline server/caas/api.js]: Don't return raw API error

File: server/caas/api.js (hunk 7 above)
Comment (westlywright) [inline server/caas/api.js]: Request has a secure method but this is the response and it just takes an object with secure property. Unless i'm misunderstanding you.

Comment (vincent99) [inline server/caas/api.js]: [req.secure](https://expressjs.com/en/4x/api.html#req.secure) is a property that tells you if the request came in via SSL.  You only want the `secure` property on the cookie fi the request is also secure.

Comment (westlywright) [inline server/caas/api.js]: 👍 my bad

File: translations/en-us.yaml [hunk 13]
```diff
@@ -2875,6 +2922,12 @@ newExternalService:
   header: Add External Service
   alert: "Note: This will only be used if the external service is a target of a load balancer."
 
+newPassword:
+  placeholder: New Password
+  password: New Password*
```
Comment (vincent99) [inline translations/en-us.yaml]: * in the template, not the translation (so we don't have to update every language with whether a field is required or not)

**Pull request merged**